### PR TITLE
transfer: support read data from stream

### PR DESCRIPTION
**Pull request by xiongjiwei**
when writing data to `storage` it needs the `size` information, it is impossible to a stream data source. This PR use `go-stream` to support reading from stream source directly.

### CODE REVIEW

**Comment by xiongjiwei** [comment]
hi @Xuanwo, I do not know what is the best practice of `go-stream`, could you please give me some advice?

**Comment by xiongjiwei** [comment]
seems we need a function like `stream.Sync()` to make sure the upper storage's data has been write into the under storage before the program exit.

File: utils/storager.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/beyondstorage/go-storage/v4/types"
+	"github.com/beyondstorage/go-stream"
+)
+
+const (
+	upperStorageConnString = "memory://"
+)
+
+var (
+	streamMap    map[string]*stream.Stream
```
Comment (Xuanwo) [inline utils/storager.go]: `stream` is designed for use concurrently, so we don't need to maintain multiple streams. Init one stream while the ftp server started is enough.

File: client/files.go [hunk 2]
```diff
@@ -66,42 +68,53 @@ func (c *Handler) upload(path string, tr utils.Conn, append bool) error {
...
 
-func (c *Handler) cacheFile(tr utils.Conn) (io.Reader, int64, error) {
+func writeToObject(ctx context.Context, label, path string, r io.Reader, storager types.Storager, o *types.Object) error {
+	b := utils.Branch(label, path)
+	if b != nil {
+		_, err := b.ReadFrom(r)
+		if err != nil {
+			return err
+		}
+		return b.Complete()
```
Comment (Xuanwo) [inline client/files.go]: `Complete()` will wait and sync all I/O operations,  just like a `Sync()`.

File: client/files.go [hunk 3]
```diff
@@ -66,42 +68,53 @@ func (c *Handler) upload(path string, tr utils.Conn, append bool) error {
...
+
 	file := new(bytes.Buffer)
-	size, err := io.Copy(file, tr)
+	size, err := io.Copy(file, r)
 	if err != nil {
-		return nil, 0, err
+		return err
+	}
+	switch label {
+	case stream.PersistMethodAppend:
```
Comment (Xuanwo) [inline client/files.go]: All data will be consumed by `stream`, so we don't need to handle `append` or write anymore.

File: utils/storager.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	if s, ok := streamMap[label]; ok {
+		b, err := s.StartBranch(atomic.AddUint64(&branchId, 1), path)
+		if err == nil {
+			return b
+		}
+	}
+	return nil
+}
+
+func StartStream(under types.Storager) {
```
Comment (Xuanwo) [inline utils/storager.go:67]: Maybe only support one kind of `stream` will make the logic easier for us?

**Comment by Xuanwo** [comment]
`go-stream` indeed lack of documents, it's ok to host a quick meeting in https://matrix.to/#/#beyondstorage@beyond-ftp:matrix.org, just ping me there.

<blockquote><div><strong><a href="https://matrix.to/#/#beyondstorage@beyond-ftp:matrix.org">You're invited to talk on Matrix</a></strong></div><div>You're invited to talk on Matrix</div></blockquote>

**Pull request merged**